src/expr/src/linear.rs

```diff
@@ -817,21 +817,28 @@ impl MapFilterProject {
     /// The supplied `shuffle` might not list columns that are not "demanded" by the
     /// instance, and so we should ensure that `self` is optimized to not reference
     /// columns that are not demanded.
-    pub fn permute(&mut self, mut shuffle: BTreeMap<usize, usize>, new_input_arity: usize) {
+    pub fn permute_fn<F>(&mut self, remap: F, new_input_arity: usize)
```

Comment:
Stale doc comment.

Reply:
Most calls to `permute_fn` do the exact same dance of creating a closure that indexes into a `Vec`. We could create a wrapper function for `permute_fn` that expects a `Vec`. (A few other call sites create a similar closure, but index into a `BTreeMap`. Some of these might be easily changed to a `Vec`, some might keep calling `permute_fn`.)

(This could be a follow-up PR, if we'd like to just merge this one.)

Reply:
I'm flexible myself, but having the closure option is important, especially in contrast to `permute` which required an owned `BTreeMap`. Tbh, "permute" is the wrong name anyhow, and it should probably be either "project" or just "visit_columns". If we end up with `project` then it is pretty idiomatic to say that it gets a `&[usize]`, because that's what we use in other project contexts.

Reply:
(Ok, I certainly wouldn't block the merging of the PR on this, as mentioned above.)